Fix tests

FreddieAkeroyd opened the pull request:
* add locking for if state machine and ioc try to use same variables and get inconsistent combination
* make sure state machine is reset to initial state at start of each test

FreddieAkeroyd committed:
Fix tests

FreddieAkeroyd committed:
Increase timeout for disconnect test

FreddieAkeroyd committed:
Add comments

FreddieAkeroyd committed:
Add comments

esmith1729 commented:
For some reason, a lot of the tests are not being skipped by the "skip_if_recsim" decorator: 

![image](https://github.com/ISISComputingGroup/EPICS-Aldn1000/assets/131008854/f6381f54-4bc2-4218-9105-ca63334ac5d0)
![image](https://github.com/ISISComputingGroup/EPICS-Aldn1000/assets/131008854/e32ec613-b17d-4013-9541-1e3a2bc53ea3)

FreddieAkeroyd committed:
Fix recsim

FreddieAkeroyd commented:
So the underlying issue was I had added a wait for a STATUS PV in tests setUp() and this PV was not being initialised in recsim. However it seems that ewith @skip_if_recsim the setUp() method is still called for the test prior to it being skipped, so if there is an error in setUp() it will look like the test is failing because it was not skipped but actually it has failed before it got to the point about deciding this due to a setUp() error. Also setUp() was starting a lewis emulator even in recsim which it does not need to do, so i have changed the setUp() method to check for recsim and act accordingly.

esmith1729 commented:
Interesting problem, so now `setUp` won't start the lewis emulator until after it checks for `uses_recsim` ?

Also this is more of a side note, but are there likely other emulators & system test suites that have the same behaviour of starting the lewis emulator, even if using RECSIM?

esmith1729 commented:
All tests pass or are now skipped as appropriate. 

Should we write a test that checks that the lock is working as intended? Basically, a test which tries to create the previous behaviour which was causing the test to fail? And then assert that it does not fail, because we have the new lock method in place?

esmith1729 committed:
Added system test to ensure that new behaviour works as intended

FreddieAkeroyd commented:
This may not be easy to do as the lack of locking was not the main problem with the test failing, and on its own may not actually lead to any failures.  I believe the main issue was the emulator being left in the wrong state at the end of a previous test which originally looked like a lack of locking. You could remove the @lockmethod decorator from the methods and see. If it doesn't fail quickly, testing may be a bit hard to do sensibly.

FreddieAkeroyd committed:
Fix decorator call

FreddieAkeroyd commented:
@esmith1729 the test added in f312c4c97ec556a2c2c28eebf3a9458091916d8d is incorrect, it only looks like it was working because the decorator was missing an argument. It seems if you have  just `@skip_if_recsim`  rather than `@skip_if_recsim("Requires emulator")` the the test never gets executed and a pass is always recorded (this is a bug). I have pushed a fix for the decorator and now the test fails - the default state of the emulator after a reset is Stopped not Paused

FreddieAkeroyd committed:
Fix tests

FreddieAkeroyd commented:
Fixed test and merging to see if fixes system tests as end of sprint - other additions can be added as separate ticket
